### PR TITLE
[Music] Export improvements and introducing the Artist Info Folder

### PR DESCRIPTION
**Pull request by DaveTBlake**
This addresses a number of issues, both fixes and improvements, related to local artist information and artwork, and the music library export facility in general.  All interrelated so had to be done at once.

It is just one of several phased changes that intend to make use of locally held artist and album data more robust and functional, and hence reduce the number of times that Kodi requests _the same music data_ from the online data providers. If we do not make more efficient use of these free online resources then with the millions of Kodi users we run the risk over over loading their servers and either having access restricted or being blocked completely. 

Feature/Change Summary
- Support for local art and nfo files for **all** artists via new  **Artist Information Folder**  and numerous fixes to better match art and nfo with correct artist
- Export done asynchronously (not lock UI), with new dialog (rather than series of yes/no dialogs), offering export of _all_ artists and allowing greater control of what exported to where.
______________________________________________________

The major enhancement at the core of this work is the concept of an **Artist Information Folder**. Until now all local artist art and nfo file processing has depended on the artist having a unique folder that contained all music by that artist and _only_ that artist. Unfortunately that is oversimpilifcation, and music often does not fit in that structure. Examples of things that don't fit include:
- artists with music on more than one music source e.g. my folder verses the kids folder
- artists with collaboration albums e.g. classical music listing composer, conductor, orchestra as album artists, or in pop music such as "Riding with the King" by Eric Clapton & BB King.
- song artists that don't have album just feature on compliations and as guest artists on albums by others
- other artists e.g. composers, producers, DJ mixers, those invloved with the music in some way but not the credited main artist

User's efforts to arrange their music in a way that compies can produce some unexpected results, that are difficult to pin down. Since the rest of the music library is agnostic about the folder layout and naming used (using embedded metadata to identify the music), it is also an necessesary imposition.

So from this PR on users will be able to identify an artist info folder - a location where all local artist art and nfo can be found. Backwards complatible, so art and nfo located next to the music files will still be found, but from v18 onwards they will be able to put it in one clear place. This single location can also be used by addons for other artist artwork, extra fanart etc. Also, unlike addons like cdArt and Skin Helper Service, Kodi has a view of all the artists and albums in the library so can handle the situation of duplicate artist names and folder naming with ease.

A number of issues that cause wrong artist and album art and nfo files to be used have also been fixed:
- Prevent export to XML writing data for more than one artist into NFO file.
- Check that an artist or album path is found before exporting that item to an NFO file. Check GetArtistPath is successful, and clear path and log when fails, before attempting to process NFO file for artist.
- Fix import to handle artists with same name by checking for match using mbid first.
- Check when getting album path that it is unique for that album and does not contain songs fro other albums.
- When getting local album artist art during initial scan to library: only look in the folder immediately above the album folder (when it contains only 1 album), and only for the first album artist if the album is a collaboration. Avoid repeatedly processing the same artist by maintaining a set.

Also a flaw meant that art set via import from xml was not fully applied, and users had to then individually  visit each album or artist manually to actually see the art. This is fixed.

**Improve music library export facility**
First thing to note is that the music lib export facility is _not_ a backup facility, and was never intended as one. Unlike video only parts of the music library are exported, and importing the resulting output will _not_ populate an empty library. That is done by scanning embedded metatdata from the music files themselves. Instead it is a means of outputing the additional artist and album data and artwork (that may have been scraped from varied online sources at different times) in a way that can be manipulated and edited by the user, and then either reloaded or transferred to another installation. It provides the main way in which online data can be captured once locally and then reused rather than being fetched from the server again.

Export of artist data was limited to just those album artists with a unique folder containing all music below it, that had scraped online data. This PR adds support for export of info and art for *all* artists in library. Having increased the data that _could_ be exported, then it is helpful to be able to select what is exported and where it is put, hence a UI change

- Replace series of yes/no responses with a dialog box and store previous values (using CServiceBroker) for use next time.
Image of Export Lib Dialog
image
- Single file output has unique (timestamped) file name.
- Separate file (nfo) output can be to a selected location.
- Support export of info and art for *all* artists in library, (not just scraped album artists with a unique folder containing all music below it)
- Add options for a) what items are exported, b) exporting items that have not been scraped, c) to export just artwork. 

Export used to take place on the same thread as the  UI, hence a delay during export (with DB access  or file writing) would cause the UI to lock. This PR moves this to be done  asynchronously.  Looking to future improvements separating DB processing from UI, it creates a music lib job queue  equivalent to that created for video in #6406 by Montellese.  

Finally there are also some minor changes to the UI when **adding a music source**
- On adding a new music source to the library, before scanning prompt optionally enter Artist Info Folder setting. Locate folder directly, not go to settings,  as in middle of adding source.
- When scan is initiated from file view context menu the option to do full rescan of tags (allowing changed system settings and tag processing to be applied regardless of files themselves being unchanged) is offered. Unnecessary when adding a new music source to the library, or from normal "Update Library" facility. This replaces the <promptfulltagscan> advanced setting added in #11557.

### CODE REVIEW

File: xbmc/interfaces/builtins/LibraryBuiltins.cpp
Comment (Rechi) [inline xbmc/interfaces/builtins/LibraryBuiltins.cpp]: (non-void) function misses a return at the end

File: xbmc/music/infoscanner/MusicInfoScanner.cpp
Comment (Rechi) [inline xbmc/music/infoscanner/MusicInfoScanner.cpp]: comparison between signed and unsigned integer

File: xbmc/music/infoscanner/MusicInfoScanner.cpp
Comment (Rechi) [inline xbmc/music/infoscanner/MusicInfoScanner.cpp]: comparison between signed and unsigned integer

File: xbmc/video/dialogs/GUIDialogVideoInfo.cpp
Comment (Rechi) [inline xbmc/video/dialogs/GUIDialogVideoInfo.cpp]: may be used uninitialized

**Review by Rechi — DISMISSED** [review]
please fix those compiler warnings

File: xbmc/settings/dialogs/GUIDialogLibExportSettings.cpp
Comment (xhaggi) [inline xbmc/settings/dialogs/GUIDialogLibExportSettings.cpp]: Please use `const` instead of `#define`

File: xbmc/interfaces/builtins/LibraryBuiltins.cpp
Comment (Rechi) [reply xbmc/interfaces/builtins/LibraryBuiltins.cpp]: thanks for fixing
As this function only returns `-1` and `0` you could switch from int return type to bool.

File: xbmc/settings/dialogs/GUIDialogLibExportSettings.cpp
Comment (DaveTBlake) [reply xbmc/settings/dialogs/GUIDialogLibExportSettings.cpp]: @xhaggi I used `#define` to be consistent with other dialogs e.g. `CGUIDialogAudioSubtitleSettings`, `CGUIDialogNetworkSetup`, `GUIDialogPVRTimerSetting`, `CGUIDialogVideoSettings` etc. I could not find anywhere that used `const` in a similar context. Maybe all need to change, or is `#define` OK?

Comment (xhaggi) [reply xbmc/settings/dialogs/GUIDialogLibExportSettings.cpp]: We should prefer const over define for newer code. Same with cast and c++ style cast.

File: xbmc/music/MusicDatabase.cpp
Comment (xhaggi) [inline xbmc/music/MusicDatabase.cpp]: `nullptr` is preferred over `NULL`, but `if (progressDialog)` should be enough here and everywhere.

File: xbmc/video/dialogs/GUIDialogVideoInfo.cpp
Comment (xhaggi) [inline xbmc/video/dialogs/GUIDialogVideoInfo.cpp]: trailing whitespace

File: xbmc/settings/dialogs/GUIDialogLibExportSettings.cpp
Comment (xhaggi) [inline xbmc/settings/dialogs/GUIDialogLibExportSettings.cpp]: `CGUIDialogLibExportSettings::%s` and `__FUNCTION__`

File: xbmc/settings/dialogs/GUIDialogLibExportSettings.cpp
Comment (xhaggi) [inline xbmc/settings/dialogs/GUIDialogLibExportSettings.cpp]: same here

**Review by xhaggi — DISMISSED** [review]

File: xbmc/settings/dialogs/GUIDialogLibExportSettings.h
Comment (DaveTBlake) [reply xbmc/settings/dialogs/GUIDialogLibExportSettings.h]: Plan is for video use too in future. But this PR is big enough so leaving video for separate one.

File: xbmc/music/MusicDatabase.cpp
Comment (DaveTBlake) [reply xbmc/music/MusicDatabase.cpp]: I have a vague memory that there was a reason for LIMIT either in SQLite or one of the MySQL derivatives. leaving this while I check

File: xbmc/music/MusicDatabase.cpp
Comment (DaveTBlake) [reply xbmc/music/MusicDatabase.cpp]: Always a utility I miss, thank you for this

File: xbmc/music/MusicDatabase.cpp
Comment (DaveTBlake) [reply xbmc/music/MusicDatabase.cpp]: Yes, the effect of PrepareSQL and need for double up to get "%" in final query

File: xbmc/settings/dialogs/GUIDialogLibExportSettings.cpp
Comment (DaveTBlake) [reply xbmc/settings/dialogs/GUIDialogLibExportSettings.cpp]: These other dialogs were all visited recently (#12251), but `#define` not replaced? Sorry to be quibbling, but this just seems an inconsistent change to make.

File: xbmc/settings/LibExportSettings.h
Comment (DaveTBlake) [reply xbmc/settings/LibExportSettings.h]: I think a class, there are methods with potential for more once extended to video export too.

File: xbmc/settings/dialogs/GUIDialogLibExportSettings.cpp
Comment (DaveTBlake) [reply xbmc/settings/dialogs/GUIDialogLibExportSettings.cpp]: I assume what you are suggesting is replace them with

File: xbmc/music/MusicDatabase.cpp
Comment (DaveTBlake) [reply xbmc/music/MusicDatabase.cpp]: OK, seems I misremembered and was thinking about situations in MySQL where ORDER BY DESC LIMIT 1 is quicker than MAX()

File: xbmc/settings/dialogs/GUIDialogLibExportSettings.cpp
Comment (xhaggi) [reply xbmc/settings/dialogs/GUIDialogLibExportSettings.cpp]: After a second look you should move them to `CSettings` h/cpp

File: xbmc/music/MusicDatabase.cpp
Comment (xhaggi) [reply xbmc/music/MusicDatabase.cpp]: Ok

File: xbmc/settings/dialogs/GUIDialogLibExportSettings.cpp
Comment (DaveTBlake) [reply xbmc/settings/dialogs/GUIDialogLibExportSettings.cpp]: That makes some sense. I have to say that it has been far from obvious how a dialog that remembers previous values should be best implemented. I have tried to follow existing examples, with varied success.

File: xbmc/settings/LibExportSettings.h
Comment (DaveTBlake) [reply xbmc/settings/LibExportSettings.h]: Probably should be called `IsItemExported()`
I started intending to have class with private attributes, but ended up leaving attributes public for convenience. My idea is that the way the list of items to be exported is held  should be encapsulated in the class, and routines that use the class as a parameter  don't need to know.

File: xbmc/settings/dialogs/GUIDialogLibExportSettings.cpp
Comment (xhaggi) [reply xbmc/settings/dialogs/GUIDialogLibExportSettings.cpp]: Adding the export settings to to media settings -> music -> export should not be an issue if you asked me. We do that for various other settings too.

File: xbmc/settings/LibExportSettings.h
Comment (xhaggi) [reply xbmc/settings/LibExportSettings.h]: That's the reason why I ask if we really need a class, a struct should be used for a plain-object with no class functionality.

File: xbmc/settings/dialogs/GUIDialogLibExportSettings.cpp
Comment (DaveTBlake) [reply xbmc/settings/dialogs/GUIDialogLibExportSettings.cpp]: Media settings are a bit of a jumble - some handled by CServiceBroker, others written by CMediaSettings directly to XML. The previous export criteria only want displaying when you are about to export, so don't belong on the media settings screen itself. Coming up on a dialog is correct. But I will follow the PVR settings (most recent addition I think) and switch to `const` in CSettings as you suggest.

File: xbmc/settings/LibExportSettings.h
Comment (DaveTBlake) [reply xbmc/settings/LibExportSettings.h]: I think there is some (admitedly minimal) class functionality, I just need to make a better job of it. Thanks for taking the time to look at this and all the corrections.

File: xbmc/settings/dialogs/GUIDialogLibExportSettings.cpp
Comment (garbear) [inline xbmc/settings/dialogs/GUIDialogLibExportSettings.cpp]: preferred c++11 style is initialize in the header

File: xbmc/settings/LibExportSettings.h
Comment (garbear) [reply xbmc/settings/LibExportSettings.h]: I agree this should be a class. The member variables should be made private to improve encapsulation.

File: xbmc/settings/LibExportSettings.cpp
Comment (garbear) [inline xbmc/settings/LibExportSettings.cpp]: New files should carry a 2017 date

File: xbmc/settings/LibExportSettings.h
Comment (DaveTBlake) [reply xbmc/settings/LibExportSettings.h]: Have improved encapsulation

**Comment by DaveTBlake** [comment]
Thanks @Rechi, @xhaggi and @garbear for the compiler warnings and C++ style corrections, and for taking the time to review this.

The overall design has been discussed on the forum https://forum.kodi.tv/showthread.php?tid=319284 and pre-PR test builds and been user tested.  I will set off another test build to ensure that in improving code quality I have not broken anything.

File: xbmc/interfaces/builtins/LibraryBuiltins.cpp
Comment (Rechi) [reply xbmc/interfaces/builtins/LibraryBuiltins.cpp]: @DaveTBlake have you overlooked the above comment, with maybe switching return type to bool?

Comment (DaveTBlake) [reply xbmc/interfaces/builtins/LibraryBuiltins.cpp]: Not overlooked. I have a bigger question over if I can make a breaking change to the builtin interface for `ExportLibrary()` rather than have `ExportLibrary2()`at all, so this routine would go. Would change return type then potentially

**Comment by DaveTBlake** [comment]
Test build 328feae8f7 available on the mirrors 
http://mirrors.kodi.tv/test-builds/windows/win32/KodiSetup-20171010-328feae8f7-HEAD-x86.exe

**Comment by zag2me** [comment]
Tested but I don't like the new dialogue asking for location on a first scan, seems an unnecessary step(can't it in default userdata?). Will update the forum thread with other suggestions.

**Comment by DaveTBlake** [comment]
Fix bug in getting album folder name that was introduced by code tidy-up

**Comment by Rechi** [comment]
@DaveTBlake this needs a rebase

File: xbmc/interfaces/builtins/LibraryBuiltins.cpp
Comment (notspiff) [reply xbmc/interfaces/builtins/LibraryBuiltins.cpp]: Nothing crucial depends on the builtin. Just change imo we never promised a stable api. The return tyoe is however not easily changeable due to the function ptr table usage.

File: xbmc/settings/dialogs/GUIDialogLibExportSettings.h
Comment (garbear) [inline xbmc/settings/dialogs/GUIDialogLibExportSettings.h]: Looks like Visual Studio screwed up the copyright header alignment

**Review by notspiff — COMMENTED** [review]
Great work, just a few small nits. And a typo which cause real shit. Great to see this improved upon again.

File: xbmc/guilib/WindowIDs.h
Comment (DaveTBlake) [reply xbmc/guilib/WindowIDs.h]: Has not been used since before Dharma, so whenever it was used it was a long time ago and not in Git. Therefore I think this is fine but I will change if required.

File: xbmc/music/jobs/MusicLibraryExportJob.cpp
Comment (DaveTBlake) [reply xbmc/music/jobs/MusicLibraryExportJob.cpp]: Negations because the settings class only has a != operator

**Comment by DaveTBlake** [comment]
Rebased, and minor corrections made. Thanks for review @notspiff  good to have your input.

I'm hoping to merge this soonest to get it (and hopefully some additional things that depend on this that I am waiting to PR) into alpha , if that seems reasonable to you all. I will squish first if we are there.

File: xbmc/guilib/WindowIDs.h
Comment (notspiff) [reply xbmc/guilib/WindowIDs.h]: nah. then it's fine. just worried about random leftover xml files in skins, but dharma is more than sufficiently back in time.

**Comment by notspiff** [comment]
this was explicitly mentioned as something we wanted in before alpha at devcon, so imho squash and merge if the issues zag ran into in forum is addressed. code is sane and any leftover bugs can be rooted out during alphas/betas.

**Comment by DaveTBlake** [comment]
Have squished prior to merge and letting jenkins build. @xhaggi I dismissed review as actions complete and I think you are happy, hope that is OK.

**Review by xhaggi — APPROVED** [review]
Sure, I trust you

File: xbmc/music/jobs/MusicLibraryExportJob.cpp
Comment (garbear) [reply xbmc/music/jobs/MusicLibraryExportJob.cpp]: Better to add a `==` operator than play negation games, but no big deal

**Comment by DaveTBlake** [comment]
I believe this has solved Trac issues:
https://trac.kodi.tv/ticket/15294
https://trac.kodi.tv/ticket/16644
But will wait a while for user feedback before closing them

**Pull request merged**
